### PR TITLE
Major refactor to Java launcher library

### PR DESCRIPTION
**Pull request by TheKodeToad**
solonovamax helped a lot with this. Thank you!

Changes:
- Rename enums to SHOUTING_SNAKE_CASE.
- Add more exit codes.
- Switch back from java.util.Logger to a custom one, this time with better code using an enum. The reason for this is the launcher did not colourise the logs like with the old behaviour. While JUL is better in terms of OOP, it doesn't have all the necessary log levels and in the end would require more code to make the log messages look nice. The API is also not the best. Using logging frameworks would just create bloat for no reason and probably create issues.
- Split the `onesix` launcher into `standard` and `legacy` (Standard, Legacy and AbstractLauncher were introduced).
- Reintroduce MultiMC's printing the chosen launcher, but from the C++ code instead.
- Reformat the code using Eclipse. It has a good formatter, okay.
- Updated license headers.
- Split Utils into different classes.
- Use MethodHandle instead of reflection where possible and practical.
- Fix warnings instead of ignoring them. If you use the correct JDK none will be created (for the source code - there is one unrelated one).
- Removed LauncherFactory again. Sorry icelimetea, but I personally think a factory is not needed for two launcher types.
- Remove constructor in NewLaunch. Why have a field when you can have a local variable? I generally also like to avoid instances of the main class.
- Remove feeding the classpath through stdin to the Java launcher. It's already passed into the java executable with the -cp flag, and NewLaunch completely ignores it.

### CODE REVIEW

**Review by flowln — COMMENTED** [review]
this creates a few new warnings:

```bash
org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
        mcParams = params.allSafe("param", Collections.EMPTY_LIST);
                                 ^
  required: String,List<String>
  found: String,List
org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
        mcParams = params.allSafe("param", Collections.EMPTY_LIST);
                                                      ^
  required: List<String>
  found:    List
org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
        mcParams = params.allSafe("param", Collections.EMPTY_LIST);
                                 ^
  required: List<String>
  found:    List
org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
        traits = params.allSafe("traits", Collections.EMPTY_LIST);
                               ^
  required: String,List<String>
  found: String,List
org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
        traits = params.allSafe("traits", Collections.EMPTY_LIST);
                                                     ^
  required: List<String>
  found:    List
org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
        traits = params.allSafe("traits", Collections.EMPTY_LIST);
                               ^
  required: List<String>
  found:    List
```

they should probably be fixed, even if it incurs in some performance overhead. this isn't exactly a hot path anyways :3

**Comment by TheKodeToad** [comment]
> this creates a few new warnings:
> 
> ```shell
> org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
>         mcParams = params.allSafe("param", Collections.EMPTY_LIST);
>                                  ^
>   required: String,List<String>
>   found: String,List
> org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
>         mcParams = params.allSafe("param", Collections.EMPTY_LIST);
>                                                       ^
>   required: List<String>
>   found:    List
> org/prismlauncher/impl/OneSixLauncher.java:63: warning: [unchecked] unchecked conversion
>         mcParams = params.allSafe("param", Collections.EMPTY_LIST);
>                                  ^
>   required: List<String>
>   found:    List
> org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked method invocation: method allSafe in class Parameters is applied to given types
>         traits = params.allSafe("traits", Collections.EMPTY_LIST);
>                                ^
>   required: String,List<String>
>   found: String,List
> org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
>         traits = params.allSafe("traits", Collections.EMPTY_LIST);
>                                                      ^
>   required: List<String>
>   found:    List
> org/prismlauncher/impl/OneSixLauncher.java:66: warning: [unchecked] unchecked conversion
>         traits = params.allSafe("traits", Collections.EMPTY_LIST);
>                                ^
>   required: List<String>
>   found:    List
> ```
> 
> they should probably be fixed, even if it incurs in some performance overhead. this isn't exactly a hot path anyways :3

I have fixed this now. Even if the performance impact is nothing to lose sleep about (well, it is for me), it's still generally preferred to use emptyList when you want an immutable empty list. Fortunately I managed to keep using this method in a safer way.

Anyway, does anyone have any idea of what I should change with the license headers, if anything?

**Comment by txtsd** [comment]
`java.applet.Applet` is deprecated in newer Java versions. Is this required for older modloaders? And can it be replaced with whatever the modern alternative for applets is?

**Comment by TheKodeToad** [comment]
> `java.applet.Applet` is deprecated in newer Java versions. Is this required for older modloaders? And can it be replaced with whatever the modern alternative for applets is?

Old versions of Minecraft used Java applets so they could be run in web browsers. I think there is a standalone main class though. However, since there are other compatibility issues with newer versions of Java, I can't see many reasons to replace it.

**Comment by solonovamax** [comment]
Perhaps it might be better to split the applet/legacy launcher and the default launcher into separate classes?

There [already exists support for different launcher types](https://github.com/PrismLauncher/PrismLauncher/blob/develop/libraries/launcher/org/prismlauncher/LauncherFactory.java#L50-L68), so why not use it?

That way you can `@Suppress` all instances of deprecated for the entire class, while only affecting the applet code.

**Comment by solonovamax** [comment]
Also, in the line of some improvements to the java code, why not use the `ParseException` class here instead of `IllegalStateException`? https://github.com/PrismLauncher/PrismLauncher/blob/deb42d1b1e71c29f7bf385c349d9c7772e09b243/libraries/launcher/org/prismlauncher/impl/OneSixLauncher.java#L93

**Comment by TheKodeToad** [comment]
Okay, I will try to improve the code more - maybe more of a refactor. I was trying to avoid it because other projects (it was mainly one though) didn't like it when I kept adding to my pull requests.

Time to fire up an IDE that isn't totally terrible ;)

**Comment by txtsd** [comment]
> Okay, I will try to improve the code more - maybe more of a refactor. I was trying to avoid it because other projects (it was mainly one though) didn't like it when I kept adding to my pull requests.
> 
> Time to fire up an IDE that isn't totally terrible ;)

Take your time and make your changes. We won't merge it until it's deemed ready.

**Comment by solonovamax** [comment]
> Okay, I will try to improve the code more - maybe more of a refactor. I was trying to avoid it because other projects (it was mainly one though) didn't like it when I kept adding to my pull requests.

I was gonna make my own pr with a bunch of code cleanup changes later as well, bc I noticed quite a few things that could be changed

**Comment by TheKodeToad** [comment]
I think I've done enough for this pull request. I've implemented https://github.com/PrismLauncher/PrismLauncher/pull/291#issuecomment-1291395239, as well as replaced some string splitting with substringing (just personal preference, but I think it's best to avoid regex when searching for single characters - and it can have side-effects if the character appears more times than you want). Maybe I'll open one aimed at improving the legacy launcher.

Also, somebody's been here before.

image

I reverted the change only for one of the fields which is changed.

No idea why re-requesting a review removed txtsd's.

**Comment by solonovamax** [comment]
few things I just noticed:

1. The license header
   ```
   /*
    *  PolyMC - Minecraft Launcher
    *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
    *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
    *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
   ```
   should be changed for
   ```
   /*
    *  Prism Launcher - Minecraft Launcher
    *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
    *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
    *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
   ```
2. You removed `OneSixLauncher.java` from the launcher code, however, I'm pretty sure that the "onesix launcher" is referenced in many other places in the codebase. May want to change that to just reference the "launcher". (Search for "onesix" or "one six" in the codebase)


Also, I'd very much like to contribute some of my own changes to this, but I want to commit them myself, so they're under my name (As they're slightly more significant than the previous changes I suggested). Would that be possible? If so, how do you want me to do that? I could make a fork of your fork, then PR to it, so that way it can be merged into this branch.

**Comment by TheKodeToad** [comment]
> few things I just noticed:
> 
>     1. The license header
>        ```
>        /*
>         *  PolyMC - Minecraft Launcher
>         *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
>         *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
>         *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        should be changed for
>        ```
>        /*
>         *  Prism Launcher - Minecraft Launcher
>         *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
>         *  Copyright (C) 2022 Jamie Mansfield <jmansfield@cadixdev.org>
>         *  Copyright (C) 2022 TheKodeToad <TheKodeToad@proton.me>
>        ```
> 
>     2. You removed `OneSixLauncher.java` from the launcher code, however, I'm pretty sure that the "onesix launcher" is referenced in many other places in the codebase. May want to change that to just reference the "launcher". (Search for "onesix" or "one six" in the codebase)
> 
> 
> Also, I'd very much like to contribute some of my own changes to this, but I want to commit them myself, so they're under my name (As they're slightly more significant than the previous changes I suggested). Would that be possible? If so, how do you want me to do that? I could make a fork of your fork, then PR to it, so that way it can be merged into this branch.

Oh dear, this is getting confusing. I'll give you push access. I'm not sure if people normally do that, but since you're not part of the Prism Launcher organisation it would be that or a fork fork (and this repo is already a fork fork).

**Comment by solonovamax** [comment]
:+1: alright

~~I pinky promise not to `git push --force`~~

**Comment by Scrumplex** [comment]
> emptyList when you want an immutable empty list

We used to use emptyList() for everything but some lists can be modified conditionally so it should actually all be mutable.

EDIT: nvm you already saw my commit :D

File: libraries/launcher/net/minecraft/Launcher.java
Comment (solonovamax) [inline libraries/launcher/net/minecraft/Launcher.java]: We should explicitly have `super();` in the constructor. It's good practice, as otherwise there's just an implied super call. (Plus, my IDE complains when it's not there)

**Review by solonovamax — CHANGES_REQUESTED** [review]

**Comment by solonovamax** [comment]
@TheKodeToad I was going to keep `LegacyUtils`, because I wanted to move a bit of the messy logic happening in the `LegacyLauncher` into that, to clean it up.

You alright if I revert that commit later?

**Comment by solonovamax** [comment]
Also, avoid force pushes if possible, because it forces me to have to rebase/re-clone my entire branch

File: libraries/launcher/net/minecraft/Launcher.java
Comment (TheKodeToad) [reply libraries/launcher/net/minecraft/Launcher.java]: In the majority of code styles, this isn't an issue. I don't see any reason to add something that doesn't change anything, since it's already fairly apparent that the class extends something else. In my experience, IDEs don't warn about it by default. However, I'm not really that bothered either way.

**Comment by TheKodeToad** [comment]
> @TheKodeToad I was going to keep `LegacyUtils`, because I wanted to move a bit of the messy logic happening in the `LegacyLauncher` into that, to clean it up.
> 
> You alright if I revert that commit later?

It seems weird to couple a class with a utils one. I feel like it would be more logical to move some of the messy logic into other methods.

> Also, avoid force pushes if possible, because it forces me to have to rebase/re-clone my entire branch

Oops, sorry. I had to because I forgot signing off. I think I've got the hang of it now.

Edit: OMG I keep pushing and then realising my mistake. Still, it shouldn't make a difference unless you already have the commits on your local copy - rebasing is normally required even without force pushing.

**Comment by TheKodeToad** [comment]
Only now do I remember that method handles are not available in Java 7.

**Comment by solonovamax** [comment]
> Edit: OMG I keep pushing and then realising my mistake. Still, it shouldn't make a difference unless you already have the commits on your local copy - rebasing is normally required even without force pushing.

If you want to make sure to sign every single commit, add the following to your `.gitconfig`:
```
[user]
    name = [YOUR NAME HERE]
    email = [YOUR EMAIL HERE]
    signingkey = [YOUR SIGNING KEY HERE]
[commit]
    gpgsign = true
```

Then, it will sign every single one of your commits by default (so you don't need to specify `-S` on the `git commit` command.)

**Comment by TheKodeToad** [comment]
> > Edit: OMG I keep pushing and then realising my mistake. Still, it shouldn't make a difference unless you already have the commits on your local copy - rebasing is normally required even without force pushing.
> 
> If you want to make sure to sign every single commit, add the following to your `.gitconfig`:
> 
> ```
> [user]
>     name = [YOUR NAME HERE]
>     email = [YOUR EMAIL HERE]
>     signingkey = [YOUR SIGNING KEY HERE]
> [commit]
>     gpgsign = true
> ```
> 
> Then, it will sign every single one of your commits by default (so you don't need to specify `-S` on the `git commit` command.)

This is for signing, not signing off.

Anyway, I just realised not to panic because turns out MethodHandles were in Java 7 and I am stupid!

**Comment by TheKodeToad** [comment]
What is the point of using an array? Then it's pretty similar to just using String::split. I chose to use substring instead due to it being clearer in my opinion.

**Comment by solonovamax** [comment]
@TheKodeToad it's like 2x-3x faster lol

I did some profiling, and on my laptop `String#split` gets around 3k-3.5k ops/ms, whereas the `String#indexOf` based split gets ~8.5k ops/ms

I want to try and optimize that startup loop, as in my experience I've noticed prism takes a bit to launch mc once you press "launch". Though, idk if that's the java code or the cpp code (haven't profiled it)

I also want to make that code as minimal as possible. I'm probably going to remove the hash map in `LauncherFactory`, and provide some optimizations to the `BufferedReader` used in reading the params (possibly use an `InputStream` directly, instead)

Also, I would have used a `Pair` or a `Tuple`, except java doesn't have either of those in the jdk.

**Comment by TheKodeToad** [comment]
Well, creating an array has a tiny performance overhead too, but it sounds like a pretty good trade-off considering it's already faster than before.

Most of the startup time is probably from the C++ side, but I'd be interested how long it takes to get from org.prismlauncher...EntryPoint to net.minecraft...Main.

BufferedReaders afaik are pretty much the best thing to use when reading line-by-line.

**Comment by solonovamax** [comment]
> BufferedReaders afaik are pretty much the best thing to use when reading line-by-line.

iirc I did some profiling a while ago, and using `InputStream` directly can actually vastly outperform it

**Comment by TheKodeToad** [comment]
I've heard that Java enums have plenty of performance issues. There's likely some *background HashMap* that we never hear about. I feel like a C-style enum would make sense - it could be passed from the C++ code with ease.

This is how I'd do it. But the integer would have to be parsed which isn't great.

```c++
enum
{
    STANDARD, LEGACY
}

launchScript += "launcher ";

// use legacy launcher if the traits are set
if (profile->getTraits().contains("legacyLaunch") || profile->getTraits().contains("alphaLaunch"))
    launchScript += LEGACY;
else
    launchScript += STANDARD;

launchScript += "\n";
```


```java
public static final int STANDARD = 0, LEGACY = 1;

public static Launcher getLauncher(int id) {
    switch (id) {
        case STANDARD:
            return new StandardLauncher();
        case LEGACY:
            return new LegacyLauncher();
    }
}
```

If we used a raw input stream, though, you can just do in.read() - '0' to get the enum ID. This would work for 10 launchers. *Maybe this is going too far though*.

**Comment by TheKodeToad** [comment]
I did not have to suppress this many warnings. Maybe you could just tweak your IDE settings on this project? Many of these are not standard warnings and just create more warnings on most compilers which don't understand them.

File: libraries/launcher/org/prismlauncher/utils/ReflectionUtils.java
Comment (solonovamax) [inline libraries/launcher/org/prismlauncher/utils/ReflectionUtils.java]: Why was this removed?

`Applet` is a deprecated api, and the compiler will produce warnings otherwise.

File: libraries/launcher/org/prismlauncher/launcher/impl/legacy/LegacyFrame.java
Comment (solonovamax) [inline libraries/launcher/org/prismlauncher/launcher/impl/legacy/LegacyFrame.java]: Why was this removed? `Applet` is a deprecated api, so it produces compiler warnings

File: libraries/launcher/org/prismlauncher/EntryPoint.java
Comment (solonovamax) [inline libraries/launcher/org/prismlauncher/EntryPoint.java]: Why are you removing the inspection suppression comments?

**Review by solonovamax — CHANGES_REQUESTED** [review]
Compiler warnings & suppression inspection changes

**Comment by solonovamax** [comment]
If you're compiling with jdk 8 or later, it will produce warnings. However, it is configured to output jdk 7 compatible class files even if you build with jdk 8, so we should keep the warning suppressions.

**Comment by TheKodeToad** [comment]
The warnings go away if you are using Java 8 or lower. Java 9 was the version that deprecated these classes. Suppressing these warnings was just causing more warnings for me.

I thought I'd already said, but I removed the inspection comments as that's something you can tweak in your IDE settings and these do not affect `javac` which is what is actually being used for compilation.

**Comment by Scrumplex** [comment]
Changing PolyMC to Prism Launcher is fine, as that is not part of the copyright header.
Copyright does not care about projects, it cares about documents and authors.

**Comment by TheKodeToad** [comment]
> Changing PolyMC to Prism Launcher is fine, as that is not part of the copyright header. Copyright does not care about projects, it cares about documents and authors.

It's inconsistent with the C++ source files though.

**Comment by solonovamax** [comment]
> It's inconsistent with the C++ source files though.

Then the cpp source files should be changed, but that's outside the scope of this PR as we're currently focusing on the java launcher component

**Comment by Scrumplex** [comment]
> It's inconsistent with the C++ source files though.

Well. We are slowly changing it to `Prism Launcher` in all files we modify. We won't do it at once for all files, but rather just for stuff we actually modify.

**Comment by TheKodeToad** [comment]
I have managed to squash and drop a lot of commits. It should be ready for review now. The code is at a stage where to see something wrong with it you'd have to be not me.

**Review by flowln — APPROVED** [review]

File: libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java
Comment (Scrumplex) [inline libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java]: I don't like this. Don't we have another way of doing this?

File: libraries/launcher/org/prismlauncher/utils/StringUtils.java
Comment (Scrumplex) [inline libraries/launcher/org/prismlauncher/utils/StringUtils.java]: drop this file

File: libraries/launcher/org/prismlauncher/EntryPoint.java
Comment (Scrumplex) [inline libraries/launcher/org/prismlauncher/EntryPoint.java]:
```suggestion
                String[] pair = input.split(" ", 2);
```

Comment (TheKodeToad) [reply libraries/launcher/org/prismlauncher/EntryPoint.java]: This is just reverting the change that was made for performance. However, if it was changed to " " instead of "\\s+" that probably would still improve performance (since I think single-character patterns are optimised).

Let me know what you think.

File: libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java
Comment (TheKodeToad) [reply libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java]: There are alternatives, but they would likely miss some of the nuances of the stock implementation - which is fairly complex.
I can understand why this class isn't ideal, though.

File: libraries/launcher/org/prismlauncher/EntryPoint.java
Comment (Scrumplex) [reply libraries/launcher/org/prismlauncher/EntryPoint.java]: Yeah, but I don't really see why we would need to have our own function here, as we should keep the code concise. I think we have been using split for the last few years without anyone caring, so I don't think the performance gain is worth the extra code.

File: libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java
Comment (Scrumplex) [reply libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java]: From what I can tell this is only used for stacktraces. I think it would make everyone's life easier if stacktraces don't have any prefix anyway.

Comment (TheKodeToad) [reply libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java]: The prefixes are used to signal the text colour to the log viewer.
The log viewer *does* automatically convert stacktraces to red though.
Just means the fatal stacktraces won't get the black background.

Comment (Scrumplex) [reply libraries/launcher/org/prismlauncher/utils/logging/LogPrintStream.java]: IMO it's fine. A fatal stacktrace generally means that the game/launcher will quit immediately. So it should be fine

**Review by Scrumplex — APPROVED** [review]
Thanks a lot!

**Pull request merged**
